JCLOUDS-664 Azurecompute-arm ImageExtension

jtjk opened the pull request:
Azurecompute-arm ImageExtension implementation

ritazh commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeImageExtension.java:
```diff
@@ -69,29 +73,28 @@
    public ImageTemplate buildImageTemplateFromNode(String name, String id) {
-      String imageName = name.toLowerCase();
-      return new ImageTemplateBuilder.CloneImageTemplateBuilder().nodeId(id).name(imageName).build();
+      name = name.toLowerCase();
```
We should probably use local variables and assume immutable parameters. Don't modify references to an input variable.

ritazh commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeImageExtension.java:
```diff
@@ -118,26 +121,47 @@ public Image call() throws Exception {
                }
                throw new UncheckedTimeoutException("Image was not created within the time limit: "
-                       + cloneTemplate.getName());
+                     + cloneTemplate.getName());
```
`cloneTemplate` and `template` are both used here. Let's stick to the local one to avoid modifying references to an input variable.

jtjk replied:
Fixed

ritazh commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeImageExtension.java:
```diff
@@ -118,26 +121,47 @@ public Image call() throws Exception {
+
+         // This removes now all the images in this storage. At least in theory, there should be just one and if there is
+         // more, they should be copies of each other.
+         BlobHelper.deleteContainerIfExists(storage, keys.key1(), "system");
```
Is it possible to have `deleteContainerIfExists` return some type of status here to verify that the container has been deleted successfully?

nacx commented on azurecompute-arm/pom.xml:
```diff
@@ -135,6 +140,11 @@
+    <dependency>
+      <groupId>org.apache.jclouds</groupId>
+      <artifactId>jclouds-blobstore</artifactId>
+      <version>2.0.0-SNAPSHOT</version>
```
Use `${project.parent.version}` too.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java:
```diff
@@ -247,6 +247,19 @@ private void getImagesFromPublisher(String publisherName, List<VMImage> osImages
       checkAndSetImageAvailability(osImages, Sets.newHashSet(locationIds));
+
+      // list custom images
+      String offer = "";
```
No need for this variable, just pass "custom" directly to the BlobHelper call.

jtjk replied:
Fixed

nacx replied:
Remove the variable?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java:
```diff
@@ -265,12 +278,18 @@ private void checkAndSetImageAvailability(List<VMImage> images, Collection<Strin
-         String version = "1";
-         VMImage ref = VMImage.create(CUSTOM_IMAGE_PREFIX + azureGroup, CUSTOM_IMAGE_PREFIX + name, sku, version, fields[0], false);
-         return ref;
+      boolean custom = fields.length == 6 ? true : false;
```
Just `boolean custom = fields.length == 6;`?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeImageExtension.java:
```diff
@@ -118,26 +121,47 @@ public Image call() throws Exception {
    @Override
    public boolean deleteImage(String id) {
+
+      String[] fields = VMImageToImage.decodeFieldsFromUniqueId(id);
```
This will at some point become hard to maintain. Could we refactor this method to return an high level object with the fields, instead of an array?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/BlobHelper.java:
```diff
@@ -0,0 +1,102 @@
+         azureBlob.close();
+      } catch (IOException e) {
+         e.printStackTrace();
+      }
```
You could just use `Closeables2.closeQuietly(azureBlob);`

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/BlobHelper.java:
```diff
@@ -0,0 +1,102 @@
+              .credentials(storage, key)
+              .buildApi(AzureBlobClient.class);
+
+      azureBlob.deleteContainer(containerName);
```
This should be wrapped in the try/catch block.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/BlobHelper.java:
```diff
@@ -0,0 +1,102 @@
+              .credentials(storage, key)
+              .buildApi(AzureBlobClient.class);
+
+      if (azureBlob.containerExists("system"))
```
Just ` return azureBlob.containerExists("system")`; also wrap it in a try/finally block to close the context.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/BlobHelper.java:
```diff
@@ -0,0 +1,102 @@
+               String dataDisk = "";
+
+               for (BlobProperties blob : blobList) {
+                  System.out.println(blob.getName());
```
Don't do this. Consider initializing this helper class with the compute logger instead.

jtjk replied:
I don't need to log it... it was unintentionally left there :)

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/BlobHelper.java:
```diff
@@ -0,0 +1,102 @@
+         e.printStackTrace();
+      }
+      try {
+         azureBlob.close();
```
Move to a finally block for the previous try

jtjk commented:
All fixed

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeImageExtension.java:
```diff
@@ -118,26 +121,39 @@ public Image call() throws Exception {
+         if (BlobHelper.customImageExists(image.storage(), keys.key1())) {
+            return false;
+         }
+         return true;
```
[minor] Just `return !BlobHelper.customImageExists(image.storage(), keys.key1());`

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/BlobHelper.java:
```diff
@@ -0,0 +1,100 @@
+      }
+      catch (Exception e) {
+         e.printStackTrace();
+      }
```
This will not fail if the container does not exist (see [here](https://github.com/jclouds/jclouds/blob/master/providers/azureblob/src/main/java/org/jclouds/azureblob/AzureBlobClient.java#L196])). Should we better not catch any exception, leave just the try/finally and let the method fail if the container can't be deleted?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/BlobHelper.java:
```diff
@@ -0,0 +1,100 @@
+      catch (Exception e) {
+         e.printStackTrace();
+         return false;
+      }
```
The container exists method should already return false [if the container exists call returns a 404](https://github.com/jclouds/jclouds/blob/master/providers/azureblob/src/main/java/org/jclouds/azureblob/AzureBlobClient.java#L156). Worth removing the catch, as it might be caused by other errors that should be taken into acount?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/BlobHelper.java:
```diff
@@ -0,0 +1,100 @@
+         }
+      } catch (Exception e) {
+         e.printStackTrace();
+      }
```
Same here. Let the method fail on uncontrolled errors.

nacx commented:
Thanks @jtjk! Just some final comments on failure propagation.
Apart from that, now deleting an image deletes the container. I recall some discussion regarding the storage account deletion. Is this related? If not, should the storage account be now deleted when there are no images/vms in there? And is this something the adapter should take into account when deleting nodes, now that we have the blob utils class in place? (If needed this can be added in a subsequent PR).

jtjk commented:
Hi, I fixed remaining issues. The storage account deletion should now be created so that if BlobHelper.customImageExists() returns false, storage should be deleted normally. I agree that it should be now fixed as a separate PR

nacx commented:
Thanks @jtjk!

    Tests run: 5, Failures: 0, Errors: 0, Skipped: 0

All image extension live tests are in green :) Mind rebasing to the latest master and fixing the conflict so I can cleanly merge it?

nacx commented:
Pushed to master as [76d7aae6](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/76d7aae6). Many thanks @jtjk!
